### DISCUSSION TITLE
Containerized development environment for one-click onboarding experience

### DISCUSSION THREAD

**Discussion by mjyc** [prl-mushr/mushr#108] [Ideas & Feedback]
Hello, a big fan of Mushr project[^1] here.

While the Mushr project's development environment (devenv) setup process is [well-documented](https://mushr.io/tutorials/noetic_quickstart/), I thought the onboarding experience could be improved by adopting a container-based devenv to provide no/low-install experience to new developers[^2]. I've been enjoying [GitHub Codespaces](https://github.com/features/codespaces) (which leverages [Development Container](https://containers.dev/))[^3] and recently found an [example of using Codespace with a ROS 2 project (Nav2)](https://discourse.ros.org/t/repeatable-reproducible-and-now-accessible-ros-development-via-dev-containers/31398). I thought Mushr could benefit from providing a devcontainer setup so that a developer could work on a Mushr project in browser (or on laptop or robot) instantly, i.e., without following installation instructions.

Here are proposed steps and their status:
1. Set up a devcontainer
    - _2023/06/04:_ Opened https://github.com/prl-mushr/mushr/pull/109, ready for review
2. Make foxglove usable in Codespaces, i.e., support visualization
    - _2023/06/07:_  Opened https://github.com/prl-mushr/mushr/pull/110 (see [demo video](https://youtu.be/OtCwCCvPoxQ)), ready for review
3. Test this containerized devenv with a robot--if makes sense

I'd love to hear the Mushr community's feedback! For example, anyone interested in trying out? what works well/not well? any must have features? etc. Looking forward to hearing from you,

---

[^1] More generally, a big [DIY robot project enthusiast](https://mjyc.github.io/2019/06/23/social.html).
[^2] Similar to how [Code Sandbox](https://codesandbox.io/) provides noinstall experience for new frontend developers. I also said low-install experience because for local development, a new developer still need to install docker, VSCode, Dev Container extension, etc.. Btw, [ROSJects from TheContstruct](https://www.youtube.com/watch?v=cR-Ow5K7oSo) tried something similar for roboticists in the past.
[^3] Learn more about how devcontainer & Codespaces can streamline robotics project setup [here](https://navigation.ros.org/development_guides/devcontainer_docs/index.html).

**Comment by mjyc** [comment]
Hey @schmittlema @cmavrogiannis @tapomayukh, how are things? Hope you are well,

I've been working on robo-DX related projects and proposed above to improve onboarding experience of Mushr. Check out this [latest demo video](https://www.youtube.com/embed/OtCwCCvPoxQ), I believe the PRs as-is now could already be helpful to new Mushr users (ugrad capstone students?) and would love to hear your feedback

**Reply by schmittlema** [reply]
Hey Mike!
This is pretty neat and at a high level seems like a good contribution to  the project! We will take a look at the specific code details in the coming days and let you know of any questions we have. Thanks for sharing!
